### PR TITLE
Viewing rights

### PR DESCRIPTION
**Pull request by SariSaar**
This PR prepares the template for handling viewing rights, denoted by the **read** permission in the user's **effectivePermissionSet** relationship.

When a user's viewing rights have been revoked:
- SearchPage redirects to NoAccessPage
- ListingPage redirects to NoAccessPage unless the user is viewing their own listing. Reviews are not displayed.
- ProfilePage redirects to NoAccessPage unless the user is viewing their own profile. Reviews are not displayed.
- TransactionPage does not show the OrderPanel for transactions at the inquiry state of booking or purchase processes.

Screenshot 2024-10-01 at 12 21 06

## New translations

```
  "NoAccessPage.viewListings.schemaTitle": "No viewing rights",
  "NoAccessPage.viewListings.heading": "You don't have viewing rights",
  "NoAccessPage.viewListings.content": "To view this content, you need to receive viewing rights from the {marketplaceName} team.",
```

## Notes

For developers who see this before the feature is fully released in the API & Docs:

When viewing rights have been revoked for a marketplace user, the following Marketplace API endpoints will be restricted:

**Listings**
GET listings/query
GET listings/show

**Reviews**
GET reviews/query
GET reviews/show

**Timeslots**
GET timeslots/query

### CODE REVIEW

File: src/components/ListingCard/ListingCard.js
Comment (SariSaar) [inline src/components/ListingCard/ListingCard.js:155]: ListingCard is used on ProfilePage, and if we want to fetch own listings to show on that page, we need to use the ownListings endpoint which yields ownListings resources.

File: src/containers/ProfilePage/ProfilePage.js
Comment (SariSaar) [inline src/containers/ProfilePage/ProfilePage.js:257]: I'm assuming here that if the user doesn't have viewing rights i.e. can't see reviews, we want to hide the whole review component instead of showing the review headings with 0 reviews. Let me know if that assumption is correct, or if we do want to show the review headings here either way!

Screenshot 2024-10-01 at 12 13 27

File: src/util/urlHelpers.js
Comment (Gnito) [inline src/util/urlHelpers.js:26]: no-viewing-right vs no-viewing-rights?

File: src/util/userHelpers.js
Comment (Gnito) [reply src/util/userHelpers.js:193]: It might also make sense to add a link to API reference > currentUser's PermissionSet 
https://www.sharetribe.com/api-reference/marketplace.html#currentuser-permissionset

...but I'm not sure when it will get every permission mentioned there.

File: src/containers/ListingPage/ListingPage.duck.js
Comment (Gnito) [inline src/containers/ListingPage/ListingPage.duck.js:420]: You need to restrict also fetchMonthlyTimeSlots call that happens in the "then" scope.

File: src/containers/ListingPage/ListingPageCarousel.js
Comment (Gnito) [inline src/containers/ListingPage/ListingPageCarousel.js:514]: My assumption is that this should be || not &&

File: src/containers/ListingPage/ListingPageCoverPhoto.js
Comment (Gnito) [inline src/containers/ListingPage/ListingPageCoverPhoto.js:512]: Same assumption as on ListingPageCarousel (`||`)

File: src/containers/ListingPage/ListingPageCarousel.js
Comment (Gnito) [reply src/containers/ListingPage/ListingPageCarousel.js:514]: You could also add a check for "isPrivateMarketplace &&" as above.

The point is that we want to make sure that on what conditions these redirections happen. Devs (future us & customizers) should not prevent SSR on listing pages on public marketplaces.

File: src/containers/NoAccessPage/NoAccessPage.js
Comment (Gnito) [inline src/containers/NoAccessPage/NoAccessPage.js:68]: There's probably a mismatch between this and Aleksi's PR.
Perhaps you should rebase your branch on top of it? Although, Aleksi might merge his branch soonish anyway.

File: src/containers/ProfilePage/ProfilePage.duck.js
Comment (Gnito) [inline src/containers/ProfilePage/ProfilePage.duck.js:246]: Add 'active' state somehow to this sentence.

"show current user's own page for active aka approved users even if they don't have viewing rights"

File: src/containers/ProfilePage/ProfilePage.duck.js
Comment (Gnito) [inline src/containers/ProfilePage/ProfilePage.duck.js:151]: Add a filter. 

This now includes draft & pending-approval & closed listings too.
Screenshot 2024-10-01 at 13 54 35

File: src/containers/ProfilePage/ProfilePage.js
Comment (Gnito) [reply src/containers/ProfilePage/ProfilePage.js:257]: Makes sense to me 👍

File: src/containers/ProfilePage/ProfilePage.js
Comment (Gnito) [inline src/containers/ProfilePage/ProfilePage.js:303]: This is not understandable.

The original setup tries to say that if page is rendered as a "preview" variant (e.g. user is in pending-approval state), then there must be a currentUser entity (it can't be null or undefined) and no errors when the call was made.

If the site is public, then user entity is the important data.

This now looks very strange and there's a hint of circular reasoning.

How about expanding this ternary to optional chaining setup?

```js
const isDataLoaded = isPreview
  ? currentUser != null || userShowError != null
  : hasNoViewingRightsOnPrivateMarketplace
  ? currentUser != null || userShowError != null
  : user != null || userShowError != null;
```

(and this could be moved below hasNoViewingRightsOnPrivateMarketplace)

File: src/containers/TransactionPage/TransactionPage.duck.js
Comment (Gnito) [inline src/containers/TransactionPage/TransactionPage.duck.js:449]: Turn this into a helper function like listingRelationship, which is defined on line 377.

File: src/containers/ListingPage/ListingPageCarousel.js
Comment (SariSaar) [reply src/containers/ListingPage/ListingPageCarousel.js:514]: If this is hasNoViewingRights || isForbiddenError, then it shows the NoAccessPage also when users are trying to reach their own listings.

By default, the .duck.js file is defined to call ownListings endpoint for all listings when a user has restricted viewing rights, and the endpoint throws 403 for everyone else's listings but gives 200 for the current user's own listings. So the combination of noViewingRights && 403 means that the user has tried to view someone else's listings, but they are getting 403 from the ownListings endpoint.

If there's some other way to handle allowing own listings but not other folks' listings, and without fetching the listing first, I can also use that in the .duck.js file.

File: src/containers/ListingPage/ListingPage.duck.js
Comment (SariSaar) [reply src/containers/ListingPage/ListingPage.duck.js:420]: Fixed!

File: src/containers/NoAccessPage/NoAccessPage.js
Comment (SariSaar) [reply src/containers/NoAccessPage/NoAccessPage.js:133]: Rebased on the most recent main and updated to include ctaData

File: src/util/userHelpers.js
Comment (SariSaar) [reply src/util/userHelpers.js]: Fixed, link added to hasPermission function comments

File: src/containers/ProfilePage/ProfilePage.duck.js
Comment (SariSaar) [reply src/containers/ProfilePage/ProfilePage.duck.js]: Comment updated!

File: src/containers/ProfilePage/ProfilePage.duck.js
Comment (SariSaar) [reply src/containers/ProfilePage/ProfilePage.duck.js]: Fixed to only include published listings 👍

File: src/containers/ProfilePage/ProfilePage.js
Comment (SariSaar) [reply src/containers/ProfilePage/ProfilePage.js]: Updated!

File: src/containers/TransactionPage/TransactionPage.duck.js
Comment (SariSaar) [reply src/containers/TransactionPage/TransactionPage.duck.js]: Extracted to a separate helper.

File: src/containers/ProfilePage/ProfilePage.js
Comment (Gnito) [inline src/containers/ProfilePage/ProfilePage.js:338]: I started to think that this "if - else if" chain does not work correctly.

```js
  if (!isDataLoaded) {
    // return etc.
  } else if (!isPreview && isNotFoundError(userShowError)) {
    // return etc.
  } else if (!isPreview && (isUnauthorizedOnPrivateMarketplace || hasUserPendingApprovalError)) {
    // return etc.
  } else if (!isPreview && !isCurrentUser && (hasNoViewingRightsOnPrivateMarketplace || isErrorNoViewingRights(userShowError))) {
    // TODO 
    // - Is it possible to create isErrorNoViewingRights error check?
    // - SSR: should only end up here if isErrorNoViewingRights is thrown (hasNoViewingRightsOnPrivateMarketplace contains assumption of currentUser entity)
    // return etc.
  } else if (!isPreview && isForbiddenError(userShowError)) {
    // The generic 403 Forbidden needs to come after more specific 403s aka hasUserPendingApprovalError (and potential isErrorNoViewingRights that doesn't exist yet)
    // return etc.
  } else if (isPreview && mounted && !isCurrentUser) {
    // return etc.
  } else if ((isPreview || isPrivateMarketplace) && !mounted) {
    // return etc.
  }
  
  // This is rendering normal profile page (not preview for pending-approval)
  return (
```

File: src/containers/ListingPage/ListingPageCarousel.js
Comment (Gnito) [reply src/containers/ListingPage/ListingPageCarousel.js]: There is error code: permission-denied-read

Comment (SariSaar) [reply src/containers/ListingPage/ListingPageCarousel.js]: If we use the ownListings.show endpoint in ListingPage.duck.js file, the 403 does not come back as permission-denied-read, because the 403 is caused by a non-author attempting ownListings.show, and the error message is `forbidden`. If we use listings.show endpoint in ListingPage.duck.js, it gives 403 permission-denied-read on all listings, including the user's own, so the isOwnListing check cannot be applied after the SDK call.

However, since read rights only apply to private marketplaces, I could set the ListingPage.duck.js hasViewingRights condition to 

`const hasViewingRights = !isPrivateMarketplace || (isPrivateMarketplace && currentUser && hasPermissionToViewData(currentUser));`

and ListingPage[variant].js hasNoViewingRights to

`const hasNoViewingRights = isPrivateMarketplace && currentUser && !hasPermissionToViewData(currentUser);`

This way, the ownListings.show endpoint (in duck.js) and hasNoViewingRights would explicitly only get applied on a private marketplace. That's what happens in practice anyway, as Console doesn't allow revoking read rights on public marketplaces, but it's also clear for customizers.

Would something like this work?

This would address the issues you listed:

- public marketplace: SSR must work without currentUser entity => hasViewingRights check does not apply
  - The aforementioned error: redirection => this won't ever happen
  - Happy case: by-pass this redirection => this always happens
- private marketplace: CSR only => hasViewingRights check applies in duck.js, currentUser already exists
  - own listings: by-pass this redirection => always shown, whether .duck.js file uses listings.show or ownListings.show
  - someone elses listings => viewing rights checked in duck.js
      - has viewing rights: by-pass this redirection => uses listings.show endpoint, no error
      - no viewing rights: redirection => uses ownListings.show endpoint, gets 403

File: src/containers/ProfilePage/ProfilePage.js
Comment (SariSaar) [reply src/containers/ProfilePage/ProfilePage.js:338]: Does SSR apply on a private marketplace?

I tested this approach anyway now to see if it's possible to create isErrorNoViewingRights here. My initial implementation was to not dispatch showUser at all when `canFetchOwnProfileOnly`, but I tested adding the dispatch in the promises array to catch the error and use it in this condition. 

If I add `showUser` dispatch and remove the `hasNoViewingRightsOnPrivateMarketplace` check from `isDataLoaded`, (so loads the page once `user` or `userShowError` exists) the `userShowError` has time to propagate before the page loads, so that I could add an `isErrorNoViewingRights` function. (This means that the isForbiddenError check in the current implementation is unnecessary, because without the showUser dispatch the userShowError always remains null).

However, if SSR doesn't apply on private marketplaces, this would only apply client-side. It would not change the behavior, and instead it would add one extra SDK call that we only make to get the error, because the `canFetchOwnProfileOnly` condition already has different dispatches going on (own listings vs all listings).

So either way works!

File: src/containers/ListingPage/ListingPageCarousel.js
Comment (Gnito) [reply src/containers/ListingPage/ListingPageCarousel.js]: ListingPage & ProfilePage are part of SSR and because the knowledge of private vs public comes from an asset (that can be stale because assets are cached for 5 min in production), we need to consider how SSR behaves on profile, listing & search pages - even if the marketplace is turned to private mode.

If I read the code correctly, the SSR during those stale moments will fetch user on profile page and listing on listing page. I doubt we can avoid it.

> Side note: I think I made a mistake earlier, when I added fetchCurrentUser to the loadData of this page. Luckily, full page loading doesn't fetch currentUser because isAuthenticated flag is false at that point of data loading and the fetch is not done on the server.

Comment (SariSaar) [reply src/containers/ListingPage/ListingPageCarousel.js:516]: Listing pages NoAccessPage condition tested with SSR-setup (i.e. private marketplace set to false on asset and no current user) and set to

`} else if (hasNoViewingRights && isForbiddenError(showListingError)) {`

=> this condition shows NoAccessPage for missing viewing rights – if hasViewingRights resolves to false in .duck.js, the endpoint that gets used is ownListings, and the error from that endpoint is `forbidden` for no current user. Removed the privateMarketplace check so this catches SSR-situations as well.

File: src/containers/ListingPage/ListingPageCoverPhoto.js
Comment (SariSaar) [reply src/containers/ListingPage/ListingPageCoverPhoto.js:513]: Fixed similarly!

File: src/containers/ProfilePage/ProfilePage.js
Comment (SariSaar) [reply src/containers/ProfilePage/ProfilePage.js]: Condition updated to 

`  } else if (
    (!isPreview && hasNoViewingRightsOnPrivateMarketplace && !isCurrentUser) ||
    isErrorNoViewingPermission(userShowError)
  ) {`

and isErrorNoViewingPermission added.

I also added similar handling to search pages:

`  } else if ((isPrivateMarketplace && hasNoViewingRightsUser) || hasNoViewingRightsError) {`

File: src/containers/ListingPage/ListingPage.duck.js
Comment (Gnito) [inline src/containers/ListingPage/ListingPage.duck.js:418]: Sorry, one more thing.
This can't default to ownListings endpoint on server because then the server can't return correct HTTP headers on public marketplace scenario. I.e. if listing is not found from public endpoint listings.show, the 404 should be set.

Comment (SariSaar) [reply src/containers/ListingPage/ListingPage.duck.js]: Fixed ListingPage.duck.js condition to default to listings.show, and updated the listing pages condition accordingly

File: src/containers/TransactionPage/TransactionPage.test.js
Comment (Gnito) [inline src/containers/TransactionPage/TransactionPage.test.js:1251]: Side note: I haven't used `.toBeNull()` expectation. So, another approach is `.not.toBeInTheDocument()`

File: src/containers/ProfilePage/ProfilePage.test.js
Comment (Gnito) [inline src/containers/ProfilePage/ProfilePage.test.js:242]: There are some surprises. (Old code is missing queryReviewsRequest. It might make sense to fix it at some point but perhaps not inside this PR.)

**Review by Gnito — APPROVED** [review]
LGTM!

File: src/containers/TransactionPage/TransactionPage.test.js
Comment (SariSaar) [reply src/containers/TransactionPage/TransactionPage.test.js:1251]: I originally tried with `.not.toBeInTheDocument()`, but I kept getting errors that the test engine wasn't able to find the item, i.e. the page was behaving correctly but the test expectation wasn't catching it correctly. On the other hand, the combination of queryByText and toBeNull resolved correctly.

Comment (SariSaar) [reply src/containers/TransactionPage/TransactionPage.test.js:1251]: Looks like the issue was with screen.getByText vs screen.queryByText and not the .not.toBeInTheDocument vs .toBeNull. So I could update these to `expect(screen.queryByText(...)).not.toBeInTheDocument()`!

Comment (Gnito) [reply src/containers/TransactionPage/TransactionPage.test.js:1251]: @SariSaar I think this is fine. 
It's just more dramatic to be able to ask the big question: "toBe or not.toBe?"

**Pull request merged**
